migration/migration.go

```diff
@@ -0,0 +1,175 @@
+package migration
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	logger "github.com/ipfs/go-log/v2"
+)
+
+var (
+	log = logger.Logger("migrations")
+
+	keyCurrentVersion = datastore.NewKey("version")
+)
+
+// Migrator ensures a datastore goes through all the needed
+// Migrations to upgrade its current version to the latest version.
+type Migrator struct {
+	ds         datastore.TxnDatastore
+	migrations map[int]Migration
+}
+
+// Migration runs a vA->v(A+1) migration.
+type Migration func(datastore.Txn) error
+
+// New returns a new Migrator.
+func New(ds datastore.TxnDatastore, migrations map[int]Migration) *Migrator {
```

Comment:
Gets created with the underlying datastore, and a map of versions and `Migration`.
i.e:
`{1: functionThatMigratesFromV0ToV1, 2: functionThatMigratesFromV1ToV2, etc}`